keps/sig-cluster-lifecycle/0035-20190128-addons-via-operators.md

```diff
@@ -0,0 +1,222 @@
+* Addon manifests are able express an operator minimum version requirement, so
+  that an addon with new requirements can require that the operator be updated
+  first
+* Airgapped operation should be possible by combining a registry mirror and
+  storage of the underlying manifest in the cluster itself.
+
+
+An example can make this easier to understand, here is what a CRD instance for
+kube-proxy might look like:
+
+```yaml
+apiVersion: addons.sigs.k8s.io/v1alpha1
+kind: KubeProxy
+metadata:
+  name: default
+  namespace: kube-system
+spec:
+  clusterCidr: 100.64.0.0/10
+  version: 1.14.4
+status:
+  healthy: true
+```
+
+This particular manifest is pinned to `version: 1.14.4`.  We could also
+subscribe to a stream of updates with a field like `channel: stable`.
+
+### Risks and Mitigations
+
+This will involve running a large number of new controllers.  This will require
+more resources; we can mitigate this by combining them into a single binary
+(similar to kube-controller-manager).
+
+Automatically updating addons could result in new SPOFs, we can mitigate this
+through mirroring (including support for air-gapped mirrors).
+
+Providing a good set of addons could result in a monoculture where mistakes
+affect most/all kubernetes clusters (even if we don’t mandate adoption, if we
+succeed we hope for widespread adoption).  We can continue with our strategies
+that we use for core components such as kube-apiserver: primarily we must keep
+the notion of stable vs less-stable releases, to stagger the risk of a bad
+rollout.  We must also consider this a trade-off against the risk that without
+coordination each piece of tooling must reinvent the wheel; we expect more
+mistakes (even measured per cluster) in that scenario.
+
+## Success Criteria
+
+We will succeed if addon operators are:
+
+* Used: addon operators are adopted by the majority of cluster installation
+tooling
+* Useful: users are generally satisfied with the functionality of addon
+operators and are not trying to work around them, or making lots of proposals /
+PRs to extend them
+* Ubiquitous: the majority of components include an operator
+* Federated: the components maintain their own operators, encoding their
+knowledge of how best to run their addon.
+
+## Graduation Criteria
+
+alpha: addon-operators are used to manage kube-proxy & CoreDNS:
+* in kube-up
+* in kubeadm (at least as an option)
+* in kops
+* in cluster-api-provider-aws & cluster-api-provider-gcp
+* adoption is documented for use by other tooling / self-managed clusters
+
+(post-alpha criteria will be added post-alpha)
+
+## Implementation History
+
+[Addon Operator session](https://www.youtube.com/watch?v=LPejvfBR5_w) given by jrjohnson & justinsb at Kubecon NA - Dec 2018
+KEP created - Jan 29 2019
+
+## Infrastructure Needed
+
+Initial development of the tooling can probably take place as part of
+kubebuilder
+
+We should likely create a repo for holding the operators themselves.  Eventually
```

Comment:
It might be good to have one repo per operator as opposed to a single repo for all operators.

Reply:
I'm already concerned today with the fragmentation that exists in the CAP*, and I think this would provide yet another degree of freedom where I'd really like to bake the defaults into a single canonical location.

Reply:
I agree with @detiber  A single repo for all operators doesn't allow the owners of the addons to build their operators where and how they want. If we know of [addon operators](https://github.com/kubernetes/enhancements/pull/746/files#r266642351) that don't need to cycle exactly with a particular kubernetes release, a single repo doesn't seem like the right approach to enabling that scenario.

Define the API contract an operator must fulfill (status/condition reporting, how the operator says if it has rolled out the expected version, etc). Have tooling to make it easy for an operator to bootstrap a project to conform to that API contract.